### PR TITLE
MVC_SPEC-27 Added MvcContext#getLocale and the LocaleResolver SPI

### PR DESCRIPTION
**Pull request by chkal**
Proposal for [MVC_SPEC-27](https://java.net/jira/browse/MVC_SPEC-27) "Handling of Locale and Language".

### CODE REVIEW

File: api/src/main/java/javax/mvc/locale/LocaleResolverContext.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+ * @since 1.0
+ */
+public interface LocaleResolverContext {
+
+    /**
+     * Returns HTTP request object from the Servlet container.
+     *
+     * @return HTTP request object.
+     */
+    HttpServletRequest getRequest();
```
Comment (ivargrimstad) [inline api/src/main/java/javax/mvc/locale/LocaleResolverContext.java]: Just a thought...
I think from what I can read from the Ozark implementation, that the only thing you really need from the request is the "Accept-Language" header, is that right?

Would it then be possible to use javax.ws.rs.core.HttpHeaders here rather than the entire request, hence get rid of the dependency on Servlet?

Comment (lefloh) [inline api/src/main/java/javax/mvc/locale/LocaleResolverContext.java]: Think the idea is that an app developer who implements a custom `LocaleResolver`could use whatever he wants from the request like a subdomain or a query parameter. We would need at least the `UriInfo` and maybe the `CookieParam`s if the servlet dependency really is an issue.

Comment (ivargrimstad) [inline api/src/main/java/javax/mvc/locale/LocaleResolverContext.java]: Good point!
What about this then:
If `HttpServletRequest` is replaced with `ContainerRequestContext` , all this information should be available and we are back to JAX-RS only (for this case)

> We can probably do the same for the ViewEngineContext, but that is another story

Comment (lefloh) [inline api/src/main/java/javax/mvc/locale/LocaleResolverContext.java]: This makes sense!

Comment (chkal) [inline api/src/main/java/javax/mvc/locale/LocaleResolverContext.java]: Hey, thanks for the feedback. My comments:

**HttpHeaders for Accept-Language**

This would possible. BUT the header doesn't just contain a locale identifier but something like this:

    Accept-Language: de-DE,de;q=0.8,en-US;q=0.6,en;q=0.4

IIRC @spericas mentioned that there is some class called `VariantSelector` in Jersey which simplifies parsing that. For now I just used `HttpServletRequest#getLocale()` because it is simpler. But I agree that the default resolver could be improved later.

**LocaleResolverContext vs ContainerRequestContext**

I agree that `LocaleResolverContext` currently doesn't contain everything an implementation could need. But we can add more things to `LocaleResolverContext` very easily. That's the nice thing about these `*Context` interfaces. They can be extended without breaking compatibility. :)

To be honest, I don't think we should use `ContainerRequestContext` instead of `LocaleResolverContext` in the `LocaleResolver` API:

  * `ContainerRequestContext` has been defined for use with `ContainerRequestFilter` which is also mentioned in the corresponding javadocs. Using it for other SPIs may be confusing for users.
  * Using `LocaleResolverContext` gives us more flexibility because we can extend this interface in later versions with MVC-specific information without breaking anything.
  * Currently Ozark calls the `LocaleResolver` from a pre-matching container filter. This means that not all the information in the `ContainerRequestContext` is available. I would prefer to make it more explicit which information is available by just including things that are available in `LocaleResolverContext`.

However, I agree that we should add things like `UriInfo` and other useful objects to `LocaleResolverContext`. And we could even remove `HttpServletRequest` if we would like to get rid of servlet dependencies and prefer JAX-RS classes. But that's something we should discuss separately.

Comment (ivargrimstad) [inline api/src/main/java/javax/mvc/locale/LocaleResolverContext.java]: I don't think I ever suggested to replace the LocaleResolverContext with anything. On the contrary, I agree with all your arguments for why it is a good thing to have.

What I **did suggest** was to get rid of the `HttpServletRequest`. And my suggestion was to replace it with the ContainerRequestContext, but as you say this may not contain the information we are after.

Maybe the `javax.ws.rs.core.Request` is possible to use instead for `HttpServletRequest`? It is available from the `ContainerRequestContext`. The `selectVariant(...)` method may be something we can use. I will see if I can manage to try this out and hopefully get back with something later today.

I really think we should try to stay within the JAX-RS domain if possible.

Comment (chkal) [inline api/src/main/java/javax/mvc/locale/LocaleResolverContext.java]: > I don't think I ever suggested to replace the LocaleResolverContext with anything. On the contrary, I agree with all your arguments for why it is a good thing to have.

Oh, sorry. I misunderstood your previous comment. I guess I should not reply so early in the morning before I had my first coffee. Sorry about that. ;-)

> Maybe the javax.ws.rs.core.Request is possible to use instead for HttpServletRequest? It is available from the ContainerRequestContext. The selectVariant(...) method may be something we can use. I will see if I can manage to try this out and hopefully get back with something later today.

Sounds like a good idea. We could try to replace `HttpServletRequest` with `Request` and also add `UriInfo` and other stuff. That should work fine.

Should we perhaps start a new discussion on the EG list whether we should prefer JAX-RS over Servlet classes? I'm just asking because AFAIK this hasn't been discussed before.

Comment (chkal) [inline api/src/main/java/javax/mvc/locale/LocaleResolverContext.java]: BTW: Even if we don't provide `HttpServletRequest` via `LocaleResolverContext`, implementation can still use `@Inject` to get it.

Comment (ivargrimstad) [inline api/src/main/java/javax/mvc/locale/LocaleResolverContext.java]: As long as it is injected in Ozark and not in any of the api code it is fine since we don't mandate the use of `HttpServletRequest`, that is left as a choice by the implementer.

Comment (chkal) [inline api/src/main/java/javax/mvc/locale/LocaleResolverContext.java]: Exactly. That's my point. So perhaps we should remove `HttpServletRequest` and add all the other stuff we get in case of a pre-matching filter from `ContainerRequestContext` (especially `UriInfo`, `Request`, etc)? If implementations need `HttpServletRequest`, they can inject it and we don't have the `HttpServletRequest` dependency in the API.

Comment (ivargrimstad) [inline api/src/main/java/javax/mvc/locale/LocaleResolverContext.java]: Now we're talking :)

Comment (chkal) [inline api/src/main/java/javax/mvc/locale/LocaleResolverContext.java]: But I still think we should discuss whether or not we want to get rid of Servlet dependencies in the API or not. Especially because the ViewEngineContext already provides a HttpServletRequest.

Comment (chkal) [inline api/src/main/java/javax/mvc/locale/LocaleResolverContext.java]: I was thinking about discussing this on the EG list in a separate thread. Thoughts?

Comment (ivargrimstad) [inline api/src/main/java/javax/mvc/locale/LocaleResolverContext.java]: Probably a good idea. I  haven't had a look yet as to how the `HttpServletRequest` is actually used from the `ViewEngineContext`, but it may very well be just as simple to remove as it was in this case.

Anyway, it would be nice to have an EG decision for a strategy when similar cases occur.

File: api/src/main/java/javax/mvc/locale/LocaleResolver.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+     *
+     * <p>If the implementation is able to resolve the locale for the request, the corresponding
+     * locale must be returned. If the implementation cannot resolve the locale, it must return
+     * <code>null</code>. In this case the resolving process will continue with the next
+     * resolver.</p>
+     *
+     * @param context the context needed for processing.
+     * @return The resolved locale or <code>null</code>.
+     */
+    Locale resolveLocale(LocaleResolverContext context);
```
Comment (ivargrimstad) [inline api/src/main/java/javax/mvc/locale/LocaleResolver.java]: What do you think? This should return an `Optional<Locale>`?

Comment (chkal) [inline api/src/main/java/javax/mvc/locale/LocaleResolver.java]: Not sure. Actually I don't know about any of the EE APIs which use `Optional`. This is rather uncommon I think. 

`Optional` is very nice if you want to explicitly tell the caller that there may be "no result". But in this case the caller is the MVC implementation, and we know what the method returns. So I'm -1 for using `Optional`.

Anyone else has thoughts regarding this?

Comment (ivargrimstad) [inline api/src/main/java/javax/mvc/locale/LocaleResolver.java]: The reason why it hasn't been used yet in the Java EE apis is that it is new with Java SE 8. And one of the goals for Java EE 8 is to make use Jave SE 8 features.

Since this is an "internal" method only used by the LocaleResolverChain at the moment, it is not a big deal. But it would make the implementation easier, clearer and more fluent. Get rid of the null checks...

Comment (lefloh) [inline api/src/main/java/javax/mvc/locale/LocaleResolver.java]: I like `Optional`s when an alternative is used but in this case we'll take the result of the next resolver. So 0 from me.

Comment (chkal) [inline api/src/main/java/javax/mvc/locale/LocaleResolver.java]: Still not sure I like using `Optional`. But as I said, this may be a matter of taste.

In my opinion the main benefit of the `Optional` class is to make the possible result of a method invocation clearer for the **caller**. But IMO the implementation gets a bit harder as you have to return  `Optional.empty()` instead of simply `null`.

Also using `Optional<Locale>` as the return type creates some interesting corner cases. The implementation could still return `null` for some reason. The compiler won't help you here. So actually the `LocaleResolverChain` will have to check for `null` **and** `isPresent()` making our implementation even more verbose.

So to sum it up. I personally like `Optional` for method the developer calls. But IMO it creates more complexity if the developer has to implement a method that returns `Optional`.

However, as I already said. Perhaps a matter of taste? :) I really would like to hear more options about this. But it doesn't like the other experts want to join us. :(

File: api/src/main/java/javax/mvc/MvcContext.java [hunk 3]
```diff
@@ -123,4 +125,12 @@
      * @return instance of encoders.
      */
     Encoders getEncoders();
+
+    /**
+     * Returns the locale of the current request.
+     *
+     * @return The request locale
+     */
```
Comment (Turini) [inline api/src/main/java/javax/mvc/MvcContext.java]: I'd keep it in a single line to avoid ambiguity in docs. I.e:
```
/**
 * @return the locale of the current request.
*/
```
What do you think?

File: api/src/main/java/javax/mvc/locale/LocaleResolver.java (hunk 2 above)
Comment (Turini) [inline api/src/main/java/javax/mvc/locale/LocaleResolver.java]: > The implementation could still return null for some reason. The compiler won't help you here.

We should always return `Optional.ofNullable()` for nullable objects, so it will never be null, right?

our implementation could move from:
```
for (LocaleResolver candidate : candidates) {
      Locale locale = candidate.resolveLocale(context);
      if (locale != null) {
         return locale;
      }
}

throw new IllegalStateException("Could not resolve locale ...");
```
to a bit more functional style:
```
candidates.stream()
    .map(c -> c.resolveLocale(context))
    .findFirst()
    .orElseThrow(() -> new IllegalStateException("Could not resolve locale ..."));
}
```
I'm ok using null here, but IMHO we should use as many as possible the jdk8 possibilities

Comment (lefloh) [inline api/src/main/java/javax/mvc/locale/LocaleResolver.java]: But you don't know what an app developer might return... 
However a +1 for the functional style :)

Comment (ivargrimstad) [inline api/src/main/java/javax/mvc/locale/LocaleResolver.java]: If the method returns an Optional and the JavaDoc states that it should do so, then the developer that returns null breaks the contract. Nothing we can prevent, nor should. 
Same goes with `return List<Foo>`... you never now if it is null, but you have to trust it is at least an empty list...

File: api/src/main/java/javax/mvc/MvcContext.java (hunk 3 above)
Comment (ivargrimstad) [inline api/src/main/java/javax/mvc/MvcContext.java]: I think you need the first sentence with a dot (.) at the end to avoid the method having empty description in higher level docs (class level).

Comment (chkal) [inline api/src/main/java/javax/mvc/MvcContext.java]: Yeah, I think Ivar is correct. Without that sentence you will get an empty text in the method table in the javadocs. So I would prefer to keep it.

File: api/src/main/java/javax/mvc/locale/LocaleResolver.java (hunk 2 above)
Comment (chkal) [inline api/src/main/java/javax/mvc/locale/LocaleResolver.java]: > We should always return Optional.ofNullable() for nullable objects, so it will never be null, right?

The interface is targeted to be implemented by application developers. So we don't know what developers may return.

> If the method returns an Optional and the JavaDoc states that it should do so, then the developer that returns null breaks the contract. Nothing we can prevent, nor should. 

I agree that the developer breaks the contract in this case. But currently there is simply no way to break the contract because there is only null and non-nul. That's what I like about the current way of doing. The developer cannot do anything wrong. By introducing `Optional` we create edge cases.

As I wrote before, not sure if using `Optional` for methods the user has to implement is a good idea. But I like `Optional` for methods that the user calls. IMO this is a huge difference.

However, if the majority of you prefers `Optional`, we can of cause change it. Just trying to make my point clear. ;)

**Comment by chkal** [comment]
Hey all! Thanks a lot for your review and feedback. I really appreciate it!

I would love to merge & close this one soon. 

There were some aspects you had different opinions about:

 * Use `Optional<Locale>` in the `LocaleResolver` or not
 * What do include in the `LocaleResolverContext`
 * Whether we need `LocaleResolverContext` at all.

Some things, like what to include in `LocaleResolverContext`, were already addressed and the pull request was updated.

So if there are still aspects you don't like about the current state of the pull request, please post a mail to the corresponding thread on the EG list. Then we can discuss it there and hopefully get a bit more feedback from the other experts.

Thanks a lot! :+1:
